spec/compiler/macro/macro_methods_spec.cr

```diff
@@ -591,6 +591,11 @@ module Crystal
         assert_macro %({{"hello world".titleize}}), %("Hello World")
       end
 
+      it "executes to_utf16" do
+        assert_macro %({{"hello".to_utf16}}), "::Slice(UInt16).literal(104_u16, 101_u16, 108_u16, 108_u16, 111_u16, 0_u16)"
```

Comment:
The trailing `0` should be allocated, but the actual slice should be truncated by one, just like in `String#to_utf16`.

Currently, you get this mismatched behaviour:
```cr
String.from_utf16("hello".to_utf16) # =>"hello"
String.from_utf16({{"hello".to_utf16}}) # =>"hello\u0000"
```

Reply:
But can we create a const literal that includes the trailing zero, but reports its size as N-1, without involving a runtime operation over it?

Reply:
Something like `Slice(T).literal(*args, *, size : Int32)`

Reply:
> But can we create a const literal that includes the trailing zero, but reports its size as N-1, without involving a runtime operation over it?

A slice literal always involves a runtime operation because `Slice.literal` calls `Slice.new`.
The implementation of the `slice_literal` declares a global variable `$Slice:{n}` with a static array containing the payload. The constant itself is then `::Slice.new(pointerof($Slice:{n}.@buffer), N)`. For a UTF16 string with trailing zero that would need to be `::Slice.new(pointerof($Slice:{n}.@buffer), N - 1)`.

In order to produce the code for that, we would need to alter the `Slice.literal` signature.

An alternative would be to implement the UTF-16 conversion directly as a primitive, i.e. it would itself be mostly identical to `slice_literal`, except that the payload is a string and transformed via `#to_utf16` and the size substraction. This primitive could be exposed as `String.from_utf16`, i.e. similar to the macro implementation, but all done in the compiler.

Reply:
Damn, that doesn't compile: we can't force named args after a splat. I wonder what's the reason :confused: 

```
 327 | def self.literal(*args, *, size : Int32? = nil)
                               ^
Error: unexpected token: "*"
```

Reply:
Anyway, that was a :hankey: thinking. The `Slice.new` struct creation itself isn't a compile time const, only the inner buffer will be a const, so slicing the trailing zero at runtime has no effect (it may even be optimized by LLVM).